cpp/inc/bond/protocol/simple_json_reader.h

```diff
@@ -179,15 +195,190 @@ class SimpleJsonReader
+            BOOST_ASSERT(!IsChild());
+        }
+
+        explicit StreamHolder(SimpleJsonReader& parent)
+            : _input(),
+              _stream(),
+              // If we're creating a child of a child, use the "root" parent
+              // as our parent instead of having a bunch of intermediate
+              // children as parent.
+              _parent(parent._streamHolder.GetParent())
+        {
+            BOOST_ASSERT(_parent->IsParent());
+
+            BOOST_ASSERT(!IsParent());
+            BOOST_ASSERT(IsChild());
+        }
+
+        StreamHolder(const StreamHolder& that)
+            : _input(),
+              _stream(),
+              _parent()
+        {
+            _input.emplace(that.GetBuffer());
+            // When the minimum version of Boost is 1.63, this can be moved
+            // to the initializer list using boost::in_place_init
+            _stream.emplace(that.GetStream(), _input.value());
+
+            BOOST_ASSERT(IsParent());
+            BOOST_ASSERT(!IsChild());
+        }
+
+        StreamHolder(StreamHolder&& that)
+            : _input(),
+              _stream(),
+              _parent(that._parent)
+        {
+            if (that.IsParent())
+            {
+                _input = std::move(that._input);
+                _stream.emplace(that._stream.value(), _input.value());
+
+                BOOST_ASSERT(IsParent());
+                BOOST_ASSERT(!IsChild());
+            }
+            else
+            {
+                BOOST_ASSERT(!IsParent());
+                BOOST_ASSERT(IsChild());
+            }
+        }
+
+        StreamHolder& operator=(const StreamHolder& that)
+        {
+            _parent = that._parent;
+
+            if (that.IsParent())
+            {
+                _input.emplace(that._input.value());
+                _stream.emplace(that._stream.value(), _input.value());
+            }
+            else
+            {
+                _input = boost::none;
+                _stream = boost::none;
+            }
+
+            BOOST_ASSERT(IsParent() == that.IsParent());
+            BOOST_ASSERT(IsChild() == that.IsChild());
+
+            return *this;
+        }
+
+        StreamHolder& operator=(StreamHolder&& that)
+        {
+            _parent = that._parent;
+
+            if (that.IsParent())
+            {
+                _input = std::move(that._input);
+                _stream.emplace(that._stream.value(), _input.value());
+
+                BOOST_ASSERT(IsParent());
+                BOOST_ASSERT(!IsChild());
+            }
+            else
+            {
+                _input = boost::none;
+                _stream = boost::none;
+
+                BOOST_ASSERT(!IsParent());
+                BOOST_ASSERT(IsChild());
+            }
+
+            return *this;
+        }
+
+        bool IsParent() const
+        {
+            return static_cast<bool>(_input) && static_cast<bool>(_stream);
+        }
+
+        bool IsChild() const
```

Comment:
It looks weird that the code for `IsChild` does not look like `!IsParent`. Maybe just keep a single `IsParent` and check the invariant only there?

Reply:
Good suggestion.